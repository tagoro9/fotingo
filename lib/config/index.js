'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _readConfigFile = require('./read-config-file');

var _readConfigFile2 = _interopRequireDefault(_readConfigFile);

var _writeConfigFile = require('./write-config-file');

var _writeConfigFile2 = _interopRequireDefault(_writeConfigFile);

var _ramda = require('ramda');

var _ramda2 = _interopRequireDefault(_ramda);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var config = (0, _readConfigFile2['default'])({
  git: {
    remote: 'origin',
    branch: 'master'
  },
  github: {
    base: 'master'
  },
  jira: {
    status: {},
    user: {}
  }
});

var data = {
  isGithubLoggedIn: function () {
    function isGithubLoggedIn() {
      return _ramda2['default'].not(_ramda2['default'].isNil(data.get(['github', 'token'])));
    }

    return isGithubLoggedIn;
  }(),
  isJiraLoggedIn: function () {
    function isJiraLoggedIn() {
      return _ramda2['default'].not(_ramda2['default'].or(_ramda2['default'].isNil(data.get(['jira', 'user', 'password'])), _ramda2['default'].isNil(data.get(['jira', 'user', 'login']))));
    }

    return isJiraLoggedIn;
  }(),
  update: _ramda2['default'].curryN(2, function (path, value) {
    config = _ramda2['default'].set(_ramda2['default'].lensPath(path), value, config);
    (0, _writeConfigFile2['default'])(config.local ? _ramda2['default'].set(_ramda2['default'].lensPath(path), value, {}) : config, config.local);
    return value;
  }),
  get: function () {
    function get(path) {
      return _ramda2['default'].view(_ramda2['default'].lensPath(path), config);
    }

    return get;
  }()
};

exports['default'] = data;