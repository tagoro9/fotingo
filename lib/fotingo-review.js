'use strict';

var _commander = require('commander');

var _commander2 = _interopRequireDefault(_commander);

var _ramda = require('ramda');

var _ramda2 = _interopRequireDefault(_ramda);

var _util = require('./git/util');

var _config = require('./config');

var _config2 = _interopRequireDefault(_config);

var _error = require('./error');

var _init = require('./init');

var _init2 = _interopRequireDefault(_init);

var _reporter = require('./reporter');

var _reporter2 = _interopRequireDefault(_reporter);

var _util2 = require('./util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

try {
  (function () {
    _commander2['default'].option('-n, --no-branch-issue', 'Do not pick issue from branch name').option('-s, --simple', 'Do not use any issue tracker').option('-l, --label [labels]', 'Labels to add to the PR', _ramda2['default'].append, []).parse(process.argv);

    var shouldGetIssue = _ramda2['default'].partial(_ramda2['default'].both(_ramda2['default'].compose(_ramda2['default'].equals(true), _ramda2['default'].prop('branchIssue')), _ramda2['default'].compose(_ramda2['default'].not, _ramda2['default'].equals(true), _ramda2['default'].prop('simple'))), [_commander2['default']]);
    var getTotalSteps = _ramda2['default'].ifElse(_ramda2['default'].propEq('simple', true), _ramda2['default'].always(4), _ramda2['default'].ifElse(_ramda2['default'].propEq('branchIssue', false), _ramda2['default'].always(5), _ramda2['default'].always(7)));

    var _reporter$stepFactory = _reporter2['default'].stepFactory(getTotalSteps(_commander2['default'])),
        step = _reporter$stepFactory.step,
        stepCurried = _reporter$stepFactory.stepCurried;

    var stepOffset = shouldGetIssue() ? 0 : 2;
    var project = (0, _util.getProject)(process.cwd());
    step(1, 'Initializing services', 'rocket');
    (0, _init2['default'])(_config2['default'], _commander2['default']).then(function (_ref) {
      var git = _ref.git,
          github = _ref.github,
          issueTracker = _ref.issueTracker;
      return (0, _util2.wrapInPromise)(step(2, 'Pushing current branch to Github', 'arrow_up')).then(_ramda2['default'].partial(git.pushBranchToGithub, [_config2['default']])).then(_ramda2['default'].ifElse(shouldGetIssue, _ramda2['default'].compose(issueTracker.getIssue, stepCurried(4, 'Getting issue from ' + issueTracker.name, 'bug'), git.extractIssueFromCurrentBranch, stepCurried(3, 'Extracting issue from current branch', 'mag_right')), _ramda2['default'].always(undefined))).then(function (issue) {
        step(5 - stepOffset, 'Getting your commit history', 'books');
        return git.getBranchInfo(issue).then(step(6 - stepOffset, 'Creating pull request', 'speaker')).then(github.checkAndGetLabels(_config2['default'], project, _commander2['default'].label)).then(github.createPullRequest(_config2['default'], project, issue, issueTracker.issueRoot));
      }).then(_ramda2['default'].ifElse(_ramda2['default'].partial(_ramda2['default'].compose(_ramda2['default'].not, _ramda2['default'].propEq('simple', true)), [_commander2['default']]), function (_ref2) {
        var issues = _ref2.branchInfo.issues,
            pullRequest = _ref2.pullRequest;
        return _ramda2['default'].compose(function (promise) {
          return promise.then(_ramda2['default'].always((0, _util2.wrapInPromise)({ pullRequest: pullRequest })));
        }, function (promises) {
          return Promise.all(promises);
        }, _ramda2['default'].map(_ramda2['default'].composeP(_ramda2['default'].partial(issueTracker.setIssueStatus, [{ status: issueTracker.status.IN_REVIEW, comment: 'PR: ' + pullRequest.html_url }]), issueTracker.getIssue, _ramda2['default'].compose(_util2.wrapInPromise, _ramda2['default'].prop('issue')))), stepCurried(7 - stepOffset, 'Setting issues to code review on ' + issueTracker.name, 'bookmark'))(issues);
      }, _ramda2['default'].identity));
    }).then(_ramda2['default'].compose(_reporter2['default'].footer, _ramda2['default'].ifElse(_ramda2['default'].isNil, _ramda2['default'].identity, _ramda2['default'].path(['pullRequest', 'html_url']))))['catch'](_error.handleError);
  })();
} catch (e) {
  (0, _error.handleError)(e);
  _commander2['default'].help();
}