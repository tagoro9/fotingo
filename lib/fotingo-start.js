'use strict';

var _commander = require('commander');

var _commander2 = _interopRequireDefault(_commander);

var _ramda = require('ramda');

var _ramda2 = _interopRequireDefault(_ramda);

var _util = require('./issue-tracker/util');

var _config = require('./config');

var _config2 = _interopRequireDefault(_config);

var _error = require('./error');

var _init = require('./init');

var _init2 = _interopRequireDefault(_init);

var _reporter = require('./reporter');

var _reporter2 = _interopRequireDefault(_reporter);

var _util2 = require('./util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _reporter$stepFactory = _reporter2['default'].stepFactory(4),
    step = _reporter$stepFactory.step,
    stepCurried = _reporter$stepFactory.stepCurried;

var getIssueId = _ramda2['default'].compose(_util.validate, _ramda2['default'].head, _ramda2['default'].prop('args'));

try {
  (function () {
    _commander2['default'].parse(process.argv);
    var issueId = getIssueId(_commander2['default']);
    step(1, 'Initializing services', 'rocket');
    (0, _init2['default'])(_config2['default'], _commander2['default']).then(function (_ref) {
      var git = _ref.git,
          issueTracker = _ref.issueTracker;
      return issueTracker.getCurrentUser().then(stepCurried(2, 'Getting \'' + issueId + '\' from ' + issueTracker.name, 'bug')).then(function (user) {
        return issueTracker.getIssue(issueId).then(issueTracker.canWorkOnIssue(user)).then(stepCurried(3, 'Setting \'' + issueId + '\' in progress', 'bookmark')).then(issueTracker.setIssueStatus({ status: issueTracker.status.IN_PROGRESS })).then(_ramda2['default'].compose(_util2.wrapInPromise, git.createBranchName)).then(stepCurried(4, function (name) {
          return 'Creating branch \'' + name + '\'';
        }, 'tada')).then(git.createIssueBranch(_config2['default']));
      });
    }).then(_reporter2['default'].footer)['catch'](_error.handleError);
  })();
} catch (e) {
  (0, _error.handleError)(e);
  _commander2['default'].help();
}