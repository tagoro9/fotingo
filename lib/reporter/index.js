'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _readline = require('readline');

var _readline2 = _interopRequireDefault(_readline);

var _nodeEmoji = require('node-emoji');

var _nodeEmoji2 = _interopRequireDefault(_nodeEmoji);

var _ramda = require('ramda');

var _ramda2 = _interopRequireDefault(_ramda);

var _read = require('read');

var _read2 = _interopRequireDefault(_read);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var stdout = process.stdout;
var startTime = Date.now();
var isTTY = stdout.isTTY;

var clearLine = function () {
  function clearLine() {
    if (!_chalk.supportsColor) {
      return;
    }
    _readline2['default'].clearLine(stdout, 0);
    _readline2['default'].cursorTo(stdout, 0);
  }

  return clearLine;
}();

var prependEmoji = function () {
  function prependEmoji(msg, emoji) {
    if (emoji && isTTY) {
      return emoji + '  ' + msg;
    }
    return msg;
  }

  return prependEmoji;
}();

var log = function () {
  function log(msg, emojiStr) {
    clearLine();
    stdout.write((emojiStr ? prependEmoji(msg, _nodeEmoji2['default'].get(emojiStr)) : msg) + '\n');
  }

  return log;
}();

var step = _ramda2['default'].curryN(4, function (total, current, msg, emojiStr) {
  var actualMsg = prependEmoji(msg, _nodeEmoji2['default'].get(emojiStr));
  log(_chalk2['default'].grey('[' + current + '/' + total + ']') + ' ' + actualMsg + '...');
});

var stepCurried = _ramda2['default'].curryN(5, function (total, current, msg, emojiStr, args) {
  step(total, current, _ramda2['default'].ifElse(_ramda2['default'].is(Function), function (m) {
    return m(args);
  }, _ramda2['default'].identity())(msg), emojiStr);
  return args;
});

var stepCurriedP = _ramda2['default'].curryN(5, function (total, current, msg, emojiStr, args) {
  stepCurried(total, current, msg, emojiStr, args);
  return Promise.resolve(args);
});

exports['default'] = {
  stepFactory: function () {
    function stepFactory(totalSteps) {
      return {
        step: step(totalSteps),
        stepCurried: stepCurried(totalSteps),
        stepCurriedP: stepCurriedP(totalSteps)
      };
    }

    return stepFactory;
  }(),

  log: log,
  info: function () {
    function info(msg) {
      return log(_chalk2['default'].grey('info') + ' ' + msg);
    }

    return info;
  }(),
  error: function () {
    function error(msg) {
      log(_chalk2['default'].red('error') + ' ' + prependEmoji(msg, _nodeEmoji2['default'].get('boom')));
    }

    return error;
  }(),
  question: function () {
    function question(_ref) {
      var _question = _ref.question,
          _ref$password = _ref.password,
          password = _ref$password === undefined ? false : _ref$password;

      return new Promise(function (resolve, reject) {
        return (0, _read2['default'])({ silent: password, prompt: _chalk2['default'].grey('question') + ' ' + _question + ':' }, function (err, text) {
          if (err) {
            return reject(err);
          }
          return resolve(text);
        });
      });
    }

    return question;
  }(),
  footer: function () {
    function footer(artifact) {
      var totalTime = ((Date.now() - startTime) / 1000).toFixed(2);
      var msg = 'Done in ' + totalTime + 's ' + (artifact ? '=> ' + artifact : '.');
      log(prependEmoji(msg, _nodeEmoji2['default'].get('sparkles')));
    }

    return footer;
  }()
};