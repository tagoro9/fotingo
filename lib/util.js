'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.promisify = exports.wrapInPromise = exports.debugCurriedP = exports.errorCurried = exports.error = exports.debugCurried = exports.debug = undefined;

var _ramda = require('ramda');

var _ramda2 = _interopRequireDefault(_ramda);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _package = require('../package.json');

var _package2 = _interopRequireDefault(_package);

var _reporter = require('./reporter');

var _reporter2 = _interopRequireDefault(_reporter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var debug = exports.debug = _ramda2['default'].curryN(2, function (module, msg) {
  return (0, _debug2['default'])(_package2['default'].name + ':' + module)(msg);
});

var debugCurried = exports.debugCurried = _ramda2['default'].curryN(3, function (module, msg, args) {
  debug(module, msg);
  return args;
});

var error = exports.error = _ramda2['default'].ifElse(_ramda2['default'].is(Error), _ramda2['default'].compose(_reporter2['default'].error, _ramda2['default'].last, _ramda2['default'].reject(_ramda2['default'].isNil), _ramda2['default'].props(['message', 'stack'])), _ramda2['default'].compose(_reporter2['default'].error, _ramda2['default'].ifElse(_ramda2['default'].is(String), _ramda2['default'].identity, _ramda2['default'].partialRight(JSON.stringify, [null, 2]))));

var errorCurried = exports.errorCurried = _ramda2['default'].curryN(2, function (msg, args) {
  error(msg);
  return args;
});

var debugCurriedP = exports.debugCurriedP = _ramda2['default'].curryN(3, function (module, msg, args) {
  debug(module, msg);
  return Promise.resolve(args);
});

var wrapInPromise = exports.wrapInPromise = function () {
  function wrapInPromise(val) {
    return Promise.resolve(val);
  }

  return wrapInPromise;
}();

var promisify = exports.promisify = function () {
  function promisify(func) {
    return function () {
      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return new Promise(function (resolve, reject) {
        return _ramda2['default'].apply(func, [].concat(_toConsumableArray(_ramda2['default'].reject(_ramda2['default'].isNil, args)), [_ramda2['default'].ifElse(_ramda2['default'].compose(_ramda2['default'].not, _ramda2['default'].isNil, _ramda2['default'].nthArg(0)), reject, _ramda2['default'].unapply(_ramda2['default'].compose(_ramda2['default'].apply(resolve), _ramda2['default'].tail)))]));
      });
    };
  }

  return promisify;
}();