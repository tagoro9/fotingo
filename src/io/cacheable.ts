import * as Keyv from 'keyv';
import { homedir } from 'os';

const path = `sqlite://${homedir()}/.fotingo_config/cache.sqlite3`;
const keyv = new Keyv(path);

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type PromiseFunction = (...args: any[]) => Promise<any>;

type Cacheable = (
  target: object,
  propertyKey: string | symbol,
  descriptor: TypedPropertyDescriptor<PromiseFunction>,
) => TypedPropertyDescriptor<PromiseFunction>;

/**
 * Decorator that caches the output of the decorated function
 * in an external data source (SQLite DB) so it can be
 * accessed across multiple executions
 * Caching is based on the function input, the specified
 * prefix and the number of minutes the data is supposed to be
 * cached
 */
export function cacheable({
  getPrefix,
  minutes,
}: {
  getPrefix?: () => string;
  minutes: number;
}): Cacheable {
  return (
    target: object,
    propertyKey: string | symbol,
    descriptor: TypedPropertyDescriptor<PromiseFunction>,
  ): TypedPropertyDescriptor<PromiseFunction> => {
    const method = descriptor.value;
    if (typeof method !== 'function') {
      throw new TypeError(
        `@cacheable decorator can only be applied to methods not: ${typeof method}`,
      );
    }

    const cachedFn: PromiseFunction = async function(...args) {
      const prefix = getPrefix ? getPrefix.call(this) : '';
      const key = `${prefix}${target.constructor.name}_${String(propertyKey)}`;
      const cachedValue = await keyv.get(key);
      if (cachedValue) {
        return Promise.resolve(cachedValue);
      }
      const result = await method.call(this, ...args);
      await keyv.set(key, result, minutes * 60 * 1000);
      return Promise.resolve(result);
    };

    descriptor.value = cachedFn;
    return descriptor;
  };
}

// One day in minutes
export const ONE_DAY = 60 * 24;
