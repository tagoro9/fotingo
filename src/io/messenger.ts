import { boundMethod } from 'autobind-decorator';
import { Observable, Subject } from 'rxjs';
import { take } from 'rxjs/operators';

export enum MessageType {
  ERROR = 'error',
  INFO = 'info',
  REQUEST = 'request',
  STATUS = 'status',
}

export enum RequestType {
  TEXT = 'text',
  PASSWORD = 'password',
  SELECT = 'select',
}

export enum Emoji {
  PENCIL = 'pencil2',
  ARROW_UP = 'arrow_up',
  BUG = 'bug',
  MAG_RIGHT = 'mag_right',
  BOOKS = 'books',
  BOOKMARK = 'books',
  BOOM = 'boom',
  OK = 'ok_hand',
  QUESTION = 'grey_question',
  LINK = 'link',
  SHIP = 'ship',
  TADA = 'tada',
}

export interface Message {
  detail?: string;
  emoji?: Emoji;
  message: string;
  type: MessageType;
  showSpinner: boolean;
}

export interface Request extends Message {
  type: MessageType.REQUEST;
  requestType: RequestType;
}

export interface Status extends Message {
  inThread: boolean;
}

export interface SelectRequest extends Request {
  options: Array<{ label: string; value: any }>;
}

export class Messenger {
  private subject: Subject<Message | Request | Status>;
  private requestSubject: Subject<any>;

  constructor() {
    this.subject = new Subject();
    this.requestSubject = new Subject();
  }

  @boundMethod
  public emit(message: string, emoji?: Emoji): void {
    this.subject.next({ emoji, message, type: MessageType.INFO, showSpinner: true });
  }

  @boundMethod
  public inThread(value: boolean): void {
    this.subject.next({
      inThread: value,
      message: 'It does not matter',
      showSpinner: false,
      type: MessageType.STATUS,
    });
  }

  public send(data: string): void {
    this.requestSubject.next(data);
  }

  public request<T>(
    question: string,
    options: { options?: Array<{ label: string; value: T }>; hint?: string } = {},
  ): Observable<T> {
    const request$ = this.requestSubject.pipe(take(1));
    this.subject.next({
      detail: options.hint,
      message: question,
      requestType: options.options ? RequestType.SELECT : RequestType.TEXT,
      showSpinner: false,
      type: MessageType.REQUEST,
      ...(options.options ? { options: options.options } : {}),
    });
    return request$;
  }

  @boundMethod
  public error(error: Error): void {
    this.subject.error(error);
  }

  public onMessage(fn: (msg: Message) => void): void {
    this.subject.subscribe({
      next: fn,
    });
  }

  public stop(): void {
    this.subject.complete();
  }
}
