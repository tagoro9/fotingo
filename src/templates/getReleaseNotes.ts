import {
  compose,
  converge,
  groupBy,
  head,
  join,
  lensProp,
  map as rMap,
  mapObjIndexed,
  prop,
  split,
  tail,
  toPairs,
  trim,
  unapply,
  view,
  zipObj,
} from 'ramda';
import { from, Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { editVirtualFile } from 'src/io/file-util';
import { Messenger } from 'src/io/messenger';
import { parseTemplate } from 'src/io/template-util';
import { Issue, IssueType, Release } from 'src/issue-tracker/Issue';
import { ReleaseConfig, ReleaseNotes } from 'src/types';

enum RELEASE_TEMPLATE_KEYS {
  VERSION = 'version',
  FIXED_ISSUES_BY_CATEGORY = 'fixedIssuesByCategory',
  FOTINGO_BANNER = 'fotingo.banner',
  JIRA_RELEASE = 'jira.release',
}

const ISSUE_TYPE_TO_RELEASE_SECTION: { [k in IssueType]: string } = {
  [IssueType.TASK]: 'Features',
  [IssueType.SUB_TASK]: 'Features',
  [IssueType.BUG]: 'Bug fixes',
  [IssueType.STORY]: 'Features',
  [IssueType.FEATURE]: 'Features',
};

/**
 * Allow the user to edit the initial release notes
 * @param initialReleaseNotes Initial release notes
 */
async function editReleaseNotes(initialNotes: string, messenger: Messenger): Promise<string> {
  messenger.inThread(true);
  const notes = await editVirtualFile({
    extension: 'md',
    initialContent: initialNotes,
    prefix: 'fotingo-review',
  });
  messenger.inThread(false);
  return notes;
}

function getReleaseNotesFromTemplate(data: Release, releaseConfig: ReleaseConfig): string {
  return parseTemplate<RELEASE_TEMPLATE_KEYS>({
    data: {
      [RELEASE_TEMPLATE_KEYS.FIXED_ISSUES_BY_CATEGORY]: compose(
        join('\n'),
        rMap(([title, list]) => `# ${title}:\n\n${list}`),
        toPairs,
        mapObjIndexed(
          compose(
            join('\n'),
            rMap((issue: Issue) => `* [#${issue.key}](${issue.url}): ${issue.fields.summary}`),
          ),
        ),
        groupBy(compose(lens => view(lens, ISSUE_TYPE_TO_RELEASE_SECTION), lensProp, prop('type'))),
      )(data.issues),
      [RELEASE_TEMPLATE_KEYS.VERSION]: data.name,
      [RELEASE_TEMPLATE_KEYS.FOTINGO_BANNER]:
        'ðŸš€ Release created with [fotingo](https://github.com/tagoro9/fotingo)',
      [RELEASE_TEMPLATE_KEYS.JIRA_RELEASE]: data.url || '',
    },
    template: releaseConfig.template,
  });
}

/**
 * Generate a function that generates the release notes
 * for a release
 * @param releaseConfig Release configuration
 * @param messenger Messenger
 */
export function getReleaseNotes(
  releaseConfig: ReleaseConfig,
  messenger: Messenger,
  release: Release,
  useDefaults: boolean,
): Observable<ReleaseNotes> {
  const initialReleaseNotes = getReleaseNotesFromTemplate(release, releaseConfig);
  return from(
    useDefaults ? [initialReleaseNotes] : editReleaseNotes(initialReleaseNotes, messenger),
  ).pipe(
    map(
      converge(unapply(zipObj(['title', 'body'])), [
        compose<string, string, string, string[], string>(head, split('\n'), trim),
        compose<string, string, string, string[], string[], string>(
          join('\n'),
          tail,
          split('\n'),
          trim,
        ),
      ]),
    ),
  );
}
