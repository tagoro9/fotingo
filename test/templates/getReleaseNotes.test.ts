import 'jest';

import { Messenger } from 'src/io/messenger';
import { getReleaseNotes } from 'src/templates/getReleaseNotes';
import { data } from 'test/lib/data';

describe('getReleaseNotes', () => {
  test('generates the release notes from the template', async () => {
    const notes = await getReleaseNotes(
      data.createReleaseConfig(),
      (jest.fn() as unknown) as Messenger,
      data.createRelease(),
      true,
    ).toPromise();
    expect(notes).toMatchSnapshot();
  });
});
