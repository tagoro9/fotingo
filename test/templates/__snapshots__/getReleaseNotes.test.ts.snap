// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getReleaseNotes generates the release notes from the template 1`] = `
Object {
  "body": "
# Bug fixes:

* [#FOTINGO-3854](https://julianne.biz): Corporate

# Features:

* [#FOTINGO-107](http://morris.com): Legacy

See [Jira release](http://benjamin.info)

ðŸš€ Release created with [fotingo](https://github.com/tagoro9/fotingo)",
  "title": "Soft",
}
`;
